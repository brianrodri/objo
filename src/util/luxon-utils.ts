import { isString, sortBy } from "lodash";
import { DateTime, Duration, Interval } from "luxon";

/**
 * @param obj - the invalid luxon object.
 * @param message - optional message to include with the error.
 *
 * @returns an {@link Error} with debug information extracted from the "invalid" input.
 */
export function newInvalidError(obj?: DateTime<false> | Duration<false> | Interval<false>, message?: string): Error {
    const lines = isString(message) ? [message] : [];
    lines.push(
        obj ?
            `Invalid${obj.constructor.name}: ${obj.invalidReason}. ${obj.invalidExplanation}`.trim()
        :   `unspecified error`,
    );
    return new Error(lines.join("\n"));
}

/**
 * @param sorted - an array of intervals sorted by start time (primary) and end time (secondary).
 *
 * @returns array of [index inclusive, index exclusive) pairs for each sub-sequence of overlapping intervals.
 */
export function getIndexCollisions(sorted: Interval<true>[]): [number, number][] {
    const collidingRanges: [number, number][] = [];

    let startIncl = 0;
    for (let stopExcl = 2; stopExcl <= sorted.length; ++stopExcl) {
        const formerStartIncl = startIncl;
        while (startIncl < stopExcl && !sorted[startIncl].intersection(sorted[stopExcl - 1])) {
            startIncl += 1;
        }
        if (startIncl - formerStartIncl > 1) {
            collidingRanges.push([formerStartIncl, stopExcl - 1]);
        }
    }
    if (sorted.length - startIncl > 1) {
        collidingRanges.push([startIncl, sorted.length]);
    }

    return collidingRanges;
}

/**
 * Asserts that the given array of intervals have no intersections.
 *
 * @param unsorted - the array of intervals to check.
 * @param message - optional message to include with the error.
 * @throws an {@link Error} if the array has overlapping intervals.
 */
export function assertNoOverlaps(unsorted: readonly Interval<true>[], message?: string): void {
    const sorted = sortBy(unsorted, "start", "end");
    const indexCollisions = getIndexCollisions(sorted);

    if (indexCollisions.length > 0) {
        const lines = isString(message) ? [message] : [];
        lines.push(
            ...indexCollisions.map(([start, end]) => {
                const overlap = Interval.fromDateTimes(sorted[start].start, sorted[end - 1].end);
                return `\t-\tindexes between [${start}, ${end}) all overlap between ${overlap}`;
            }),
        );
        throw new Error(lines.join("\n"));
    }
}
