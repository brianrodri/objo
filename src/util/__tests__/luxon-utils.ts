import { entriesIn } from "lodash";
import { DateTime, Duration, Interval } from "luxon";
import { describe, expect, it } from "vitest";

import { assertNoOverlaps, newInvalidError } from "../luxon-utils";
import { WITH_OVERLAPPING_INTERVALS, WITHOUT_OVERLAPPING_INTERVALS } from "./luxon-utils.const";

describe(newInvalidError.name, () => {
    const reason = "user-provided reason";
    const explanation = "user-provided explanation";

    const _throw = (e: Error) => {
        throw e;
    };

    describe.each(["user-provided message", undefined])("with message=%j", (message) => {
        it.each([
            DateTime.invalid(reason, explanation),
            Duration.invalid(reason, explanation),
            Interval.invalid(reason, explanation),
        ])("should accept $constructor.name", (invalidResult) => {
            expect(() => _throw(newInvalidError(invalidResult, message))).toThrowErrorMatchingSnapshot();
        });

        it("should accept undefined", () => {
            expect(() => _throw(newInvalidError(undefined, message))).toThrowErrorMatchingSnapshot();
        });
    });
});

describe(assertNoOverlaps.name, () => {
    describe.each(["user-provided message", undefined])("with message=%j", (message) => {
        it.each(entriesIn(WITHOUT_OVERLAPPING_INTERVALS))("should accept %j", (_, intervals) => {
            expect(() => assertNoOverlaps(intervals, message)).not.toThrow();
        });

        it.each(entriesIn(WITH_OVERLAPPING_INTERVALS))("should reject %j", (_, intervals) => {
            expect(() => assertNoOverlaps(intervals, message)).toThrowErrorMatchingSnapshot();
        });
    });
});
