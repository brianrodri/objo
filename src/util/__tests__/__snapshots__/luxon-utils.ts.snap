// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`assertNoOverlaps > with message="user-provided message" > should reject "group of equal intervals" 1`] = `
[Error: user-provided message
	-	indexes between [0, 5) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-07T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message="user-provided message" > should reject "group of one interval after an overlapping group" 1`] = `
[Error: user-provided message
	-	indexes between [0, 3) all overlap between [2025-03-05T00:00:00.000Z – 2025-03-10T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message="user-provided message" > should reject "group of one interval before an overlapping group" 1`] = `
[Error: user-provided message
	-	indexes between [1, 4) all overlap between [2025-03-05T00:00:00.000Z – 2025-03-10T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message="user-provided message" > should reject "group of overlapping intervals with non-overlapping intervals in between" 1`] = `
[Error: user-provided message
	-	indexes between [0, 2) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-05T00:00:00.000Z)
	-	indexes between [3, 5) all overlap between [2025-03-21T00:00:00.000Z – 2025-03-25T00:00:00.000Z)
	-	indexes between [6, 8) all overlap between [2025-04-13T00:00:00.000Z – 2025-04-17T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message="user-provided message" > should reject "group of overlapping intervals with the same end time" 1`] = `
[Error: user-provided message
	-	indexes between [0, 3) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-06T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message="user-provided message" > should reject "group of overlapping intervals with the same start time" 1`] = `
[Error: user-provided message
	-	indexes between [0, 3) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-06T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message="user-provided message" > should reject "pair of overlapping intervals" 1`] = `
[Error: user-provided message
	-	indexes between [0, 2) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-05T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message=undefined > should reject "group of equal intervals" 1`] = `[Error: 	-	indexes between [0, 5) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-07T00:00:00.000Z)]`;

exports[`assertNoOverlaps > with message=undefined > should reject "group of one interval after an overlapping group" 1`] = `[Error: 	-	indexes between [0, 3) all overlap between [2025-03-05T00:00:00.000Z – 2025-03-10T00:00:00.000Z)]`;

exports[`assertNoOverlaps > with message=undefined > should reject "group of one interval before an overlapping group" 1`] = `[Error: 	-	indexes between [1, 4) all overlap between [2025-03-05T00:00:00.000Z – 2025-03-10T00:00:00.000Z)]`;

exports[`assertNoOverlaps > with message=undefined > should reject "group of overlapping intervals with non-overlapping intervals in between" 1`] = `
[Error: 	-	indexes between [0, 2) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-05T00:00:00.000Z)
	-	indexes between [3, 5) all overlap between [2025-03-21T00:00:00.000Z – 2025-03-25T00:00:00.000Z)
	-	indexes between [6, 8) all overlap between [2025-04-13T00:00:00.000Z – 2025-04-17T00:00:00.000Z)]
`;

exports[`assertNoOverlaps > with message=undefined > should reject "group of overlapping intervals with the same end time" 1`] = `[Error: 	-	indexes between [0, 3) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-06T00:00:00.000Z)]`;

exports[`assertNoOverlaps > with message=undefined > should reject "group of overlapping intervals with the same start time" 1`] = `[Error: 	-	indexes between [0, 3) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-06T00:00:00.000Z)]`;

exports[`assertNoOverlaps > with message=undefined > should reject "pair of overlapping intervals" 1`] = `[Error: 	-	indexes between [0, 2) all overlap between [2025-03-01T00:00:00.000Z – 2025-03-05T00:00:00.000Z)]`;

exports[`newInvalidError > with message="user-provided message" > should accept 'DateTime' 1`] = `
[Error: user-provided message
InvalidDateTime: user-provided reason. user-provided explanation]
`;

exports[`newInvalidError > with message="user-provided message" > should accept 'Duration' 1`] = `
[Error: user-provided message
InvalidDuration: user-provided reason. user-provided explanation]
`;

exports[`newInvalidError > with message="user-provided message" > should accept 'Interval' 1`] = `
[Error: user-provided message
InvalidInterval: user-provided reason. user-provided explanation]
`;

exports[`newInvalidError > with message="user-provided message" > should accept undefined 1`] = `
[Error: user-provided message
unspecified error]
`;

exports[`newInvalidError > with message=undefined > should accept 'DateTime' 1`] = `[Error: InvalidDateTime: user-provided reason. user-provided explanation]`;

exports[`newInvalidError > with message=undefined > should accept 'Duration' 1`] = `[Error: InvalidDuration: user-provided reason. user-provided explanation]`;

exports[`newInvalidError > with message=undefined > should accept 'Interval' 1`] = `[Error: InvalidInterval: user-provided reason. user-provided explanation]`;

exports[`newInvalidError > with message=undefined > should accept undefined 1`] = `[Error: unspecified error]`;
